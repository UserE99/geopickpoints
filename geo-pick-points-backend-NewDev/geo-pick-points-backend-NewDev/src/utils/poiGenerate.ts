import * as GameService from "../services/GameService";
import { GameResource, POIListResource, POIResource } from "src/Resources";
import * as POIService from "../services/POIService";
import * as POILISTSService from "../services/POIListsService"
import { Types } from "mongoose";

export async function createExampleGame() {
    try {
        // Definiere alle POIs
        const berlinPOIs: POIResource[] = [
            { name: "Alexanderplatz", lat: 52.520008, long: 13.404954, beschreibung: "Ein belebter Platz mit Fernsehturm, Geschäften und urbanem Flair.", punkte: 100 },
            { name: "Brandenburger Tor", lat: 52.516275, long: 13.377704, beschreibung: "Ein ikonisches Monument und Symbol für Geschichte und Einheit.", punkte: 200 },
            { name: "Podsdamer Platz", lat: 52.509290, long: 13.376340, beschreibung: "Ein moderner Knotenpunkt mit Architektur, Kultur und Unterhaltung.", punkte: 100 },
            { name: "Oberbaumbrücke", lat: 52.501834, long: 13.445656, beschreibung: "Eine markante Brücke mit Doppeldeck-Architektur und historischem Charme.", punkte: 100 },
            { name: "Museumsinsel", lat: 52.516260, long: 13.402480, beschreibung: "Ein einzigartiges Kulturensemble mit weltberühmten Museen.", punkte: 100 },
            { name: "Volkspark Friedrichhain", lat: 52.528730, long: 13.442284, beschreibung: "Ein weitläufiger Park mit grünen Wiesen, Hügeln und Entspannungsoasen.", punkte: 50 },
            { name: "Deutsches Technikmuseum", lat: 52.498603, long: 13.378154, beschreibung: "Ein faszinierendes Museum mit historischen Exponaten zu Technik und Ingenieurskunst.", punkte: 50 },
            { name: "Checkpoint Charlie", lat: 52.507530, long: 13.390378, beschreibung: "Ein historischer Grenzpunkt und Symbol des Kalten Krieges.", punkte: 200 },
        ];

        const bhtPOIs: POIResource[] = [
            { name: "Workout Park", lat: 52.545374, long: 13.352802, beschreibung: "Workout Park", punkte: 50 },
            { name: "Spielplatz auf dem Zeppelinplatz", lat: 52.546413, long: 13.353094, beschreibung: "Spielplatz auf dem Zeppelinplatz", punkte: 200 },
            { name: "Einfahrt", lat: 52.546101, long: 13.355068, beschreibung: "Einfahrt", punkte: 50 },
            { name: "Fahrradständer Zeppelinplatz", lat: 52.545717, long: 13.351990, beschreibung: "Fahrradständer Zeppelinplatz", punkte: 100 },
            { name: "Eingang Zeppelinplatz", lat: 52.545879, long: 13.354316, beschreibung: "Eingang Zeppelinplatz", punkte: 100 },
        ];

        const potsdamPOIs: POIResource[] = [
            { name: "Park Sanssouci", lat: 52.40340322194994, long: 13.029932869765789, beschreibung: "Park Sanssouci", punkte: 50 },
            { name: "Filmpark Babelsberg", lat: 52.38494249076188, long: 13.11788422883501, beschreibung: "Filmpark Babelsberg", punkte: 200 },
            { name: "Glienicker Brücke", lat: 52.413617910656896, long: 13.090731035582499, beschreibung: "Glienicker Brücke", punkte: 50 },
            { name: "Schloss Cecilienhof", lat: 52.41958918275413, long: 13.070846179757533, beschreibung: "Schloss Cecilienhof", punkte: 100 },
            { name: "Biosphäre Potsdam", lat: 52.41901031179306, long: 13.049338804902431, beschreibung: "Biosphäre Potsdam", punkte: 100 },
        ];

        // Erstelle alle POIs und sammle die IDs
        const createPOIs = async (pois: POIResource[]): Promise<string[]> => {
            const poiIds: string[] = [];
            for (const poi of pois) {
                const fullData = await POIService.createPOI(poi);
                poiIds.push(fullData.id!);
            }
            return poiIds;
        };

        const berlinPoiIds = await createPOIs(berlinPOIs);
        const bhtPoiIds = await createPOIs(bhtPOIs);
        const potsdamPoiIds = await createPOIs(potsdamPOIs);

        // Spiel erstellen
        const gameData: GameResource = {
            title: "Berlin Sehenswürdigkeiten",
            beschreibung: "Einige der bekanntesten Sehenswürdigkeiten in Berlin",
            poilId: [...berlinPoiIds, ...bhtPoiIds],
            maxTeam: 5,
            userId: new Types.ObjectId().toString(),
        };
        await GameService.createGame(gameData);

        const gameDataPotsdam: GameResource = {
            title: "Potsdam Sehenswürdigkeiten",
            beschreibung: "Einige der bekanntesten Sehenswürdigkeiten in Potsdam",
            poilId: [...potsdamPoiIds],
            maxTeam: 5,
            userId: new Types.ObjectId().toString(),
        };
        await GameService.createGame(gameDataPotsdam);


        //1.  POI-Liste Berlin erstellen
        const poiList: POIListResource = {
            name: "Berlin",
            poilId: [...berlinPoiIds, ...bhtPoiIds],
        };
        await POILISTSService.createPOIList(poiList);

        //2.  POI-Liste Potsdam erstellen
        const poiListPotsdam: POIListResource = {
            name: "Potsdam",
            poilId: [...potsdamPoiIds],
        };
        await POILISTSService.createPOIList(poiListPotsdam);

        console.log("Beispielspiel und POI-Liste erfolgreich erstellt.");
    } catch (error: any) {
        console.error("Fehler beim Erstellen des Beispielspiels:", error.message);
        throw error;
    }
}
