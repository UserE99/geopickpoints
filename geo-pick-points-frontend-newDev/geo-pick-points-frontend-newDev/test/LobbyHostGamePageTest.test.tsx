
import React, { act } from "react";
import { render, fireEvent, screen, waitFor, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import LobbyHostGamePage from "../src/Pages/LobbyHostGamePage";
import { Provider, useDispatch } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { useGameStatus } from "../src/utils/GameStatusContext";
import { TextEncoder, TextDecoder } from 'util';
import { joinInTeam, unjoinTeam } from '../src/actions/JoinTeamGame';
import * as WebSocketSetup from '../src/utils/WebSocketSetup';


global.fetch = jest.fn(() =>
    Promise.resolve(
        new Response(JSON.stringify([]), {
            status: 200,
            headers: { "Content-Type": "application/json" },
        })
    )
) as jest.Mock;

jest.mock("../src/actions/JoinTeamGame", () => {
    const actualModule = jest.requireActual("../src/actions/JoinTeamGame");

    return {
        ...actualModule,
        joinInTeam: Object.assign(jest.fn(), actualModule.joinInTeam),
    };
});

// Mock Redux Store
const mockStore = configureStore([]);

jest.mock("../src/utils/GameStatusContext", () => ({
    useGameStatus: jest.fn(),
}));

jest.mock("../src/utils/WebSocketSetup", () => ({
    __esModule: true,
    default: jest.fn(() => ({
        messages: [],
        sendMessage: jest.fn(),
    })),
}));

jest.mock("../src/actions/JoinTeamGame", () => ({
    joinInTeam: jest.fn(),
    unjoinTeam: jest.fn(),
}));

jest.mock("../src/actions/CreateGameInstance", () => ({
    createGameInstance: jest.fn(),
}));

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: jest.fn(),
}));
jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useDispatch: jest.fn(),
}));

describe("LobbyHostGamePage Component", () => {

    let store;
    let mockDispatch;

    beforeEach(() => {
        store = mockStore({
            teams: [],
        });
        store.dispatch = jest.fn();

        mockDispatch = jest.fn();
        const mockUseDispatch = useDispatch as jest.MockedFunction<typeof useDispatch>;
        mockUseDispatch.mockReturnValue(mockDispatch);

        (useGameStatus as jest.Mock).mockReturnValue({
            isGameStarted: false,
            setIsGameStarted: jest.fn(),
        });
    });


    it("renders LobbyHostGamePage correctly and matches snapshot", async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <LobbyHostGamePage />
                </MemoryRouter>
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("displays the correct header and team list when data is present", async () => {
        const mockDataTeam = [
            {
                _id: "team1",
                name: "Team Alpha",
                players: [{ nickName: "Player 1", id: "player1", name: "John Doe" }],
                codeInvite: "abc123",
                shareUrl: "http://example.com/invite",
                qaCode: "http://example.com/qrcode",
                playersIDData: "data",
            },
        ];

        jest.spyOn(global, "fetch").mockResolvedValueOnce({
            json: async () => mockDataTeam,
            ok: true,
        } as any);

        await act(async () => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <LobbyHostGamePage />
                    </MemoryRouter>
                </Provider>
            );
        });

        expect(screen.getByText("Your Lobby Host")).toBeInTheDocument();
        expect(screen.getByText("List of your Teams")).toBeInTheDocument();
        expect(screen.getByText("Team Alpha")).toBeInTheDocument();
    });


    it('displays an error message if fetching team data fails', async () => {
        jest.spyOn(global, 'fetch').mockRejectedValueOnce(new Error('Failed to fetch team data'));

        await act(async () => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <LobbyHostGamePage />
                    </MemoryRouter>
                </Provider>
            );
        });

        await waitFor(() => {
            expect(screen.queryByText(/No teams available/i)).toBeTruthy();
        });


        global.fetch.mockRestore();
    });

    it('handles joining and leaving a team', async () => {
        const store = mockStore({
            gameStatus: { isGameStarted: false },
            playerTeamInfo: { teamId: null, nickName: null },
        });


        // Dispatch als Mock-Funktion setzen
        store.dispatch = jest.fn();
        const mockSendMessage = jest.fn();
        (WebSocketSetup.default as jest.Mock).mockReturnValue({
            messages: [],
            sendMessage: mockSendMessage,
        });
        const mockTeamData = [
            {
                _id: 'team1',
                name: 'Team One',
                shortName: 'T1',
                players: [
                    { nickName: 'TestPlayer', playerId: '1234' },
                ],
            },
        ];

        jest.spyOn(global, 'fetch').mockResolvedValueOnce({
            ok: true,
            json: async () => mockTeamData,
        } as any);

        render(
            <Provider store={store}>
                <MemoryRouter>
                    <LobbyHostGamePage />
                </MemoryRouter>
            </Provider>
        );

        // Warte auf das Laden der Teams
        await waitFor(() => expect(screen.getByText(/List of your Teams/i)).toBeInTheDocument());

        // Verifiziere den Host-Namen
        const hostNameContainer = screen.getByText(/Host Name :/i).closest('h5');

        // "Join Team" Button klicken
        const joinButton = await screen.findByRole('button', { name: /Join Team/i });
        fireEvent.click(joinButton);

        expect(mockSendMessage).toHaveBeenCalledWith({
            playerId: undefined,
            playerName: undefined,
            teamId: "team1",
            type: "join"

        });

        // Stelle sicher, dass der "Leave Team" Button sichtbar wird
        await waitFor(() => {
            expect(screen.getByRole('button', { name: /Leave Team/i })).toBeInTheDocument();
        });

        // Simuliere Klick auf "Leave Team"
        const leaveButton = screen.getByRole('button', { name: /Leave Team/i });
        fireEvent.click(leaveButton);

        // Überprüfe, ob mockSendMessage für "leave" aufgerufen wurde
        expect(mockSendMessage).toHaveBeenCalledWith({
            type: 'leave',
            playerId: '1234',
            playerName: 'TestPlayer',
            teamId: 'team1',
        });

        // API-Mocks zurücksetzen
        global.fetch.mockRestore();
    });



    it("handles the start game functionality", async () => {
        const mockDataTeam = [
            {
                _id: "team1",
                name: "Team Alpha",
                players: [{ nickName: "TestPlayer", id: "player1", name: "John Doe" }],
                codeInvite: "abc123",
                shareUrl: "http://example.com/invite",
                qaCode: "http://example.com/qrcode",
                playersIDData: "data",
            },
        ];

        jest.spyOn(global, 'fetch').mockResolvedValueOnce({
            ok: true,
            json: async () => mockDataTeam,
        } as any);


        await act(async () => {
            render(
                <Provider store={store}>
                    <MemoryRouter>
                        <LobbyHostGamePage />
                    </MemoryRouter>
                </Provider>
            );
        });

        expect(screen.getByText("TestPlayer (Your)")).toBeInTheDocument();

        const startGameButton = screen.getByText("Start Game!");
        fireEvent.click(startGameButton);

        expect(mockDispatch).toHaveBeenCalledTimes(1);
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });
});  
