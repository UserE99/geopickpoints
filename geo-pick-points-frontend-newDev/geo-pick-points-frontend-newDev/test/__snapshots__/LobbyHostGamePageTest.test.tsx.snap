// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LobbyHostGamePage Component renders LobbyHostGamePage correctly and matches snapshot 1`] = `
<DocumentFragment>
  <h3
    style="margin: 1.5rem;"
  >
    Your Lobby Host
  </h3>
  <div
    class="container mt-5"
  >
    <div
      class="card"
    >
      <div
        class="card-header text-primary"
      >
        <h4>
          List of your Teams
        </h4>
      </div>
      <div
        class="text-center my-4"
      >
        <h5>
          Invite QR Code
        </h5>
        <div>
           
        </div>
        <div
          style="margin-top: 1rem; margin-bottom: 0rem;"
        >
          <h5
            class="fw-bold"
          >
            Host Name : 
          </h5>
        </div>
      </div>
      <div
        class="card-body row"
      >
        <p>
          No teams available
        </p>
      </div>
    </div>
    <button
      class="btn btn-danger"
      type="button"
    >
       Start Game!
    </button>
  </div>
</DocumentFragment>
`;
